src/konserve/cache.cljc

```diff
@@ -21,84 +23,130 @@
 
 (defn exists?
   "Checks whether value is in the store."
-  [store key]
-  (go-locked
-   store key
-   (or (cache/has? @(:cache store) key)
-       (<! (-exists? store key)))))
+  ([store key]
+   (exists? store key false))
+  ([store key opts]
+   (trace "exists? on key " key opts)
+   (async+sync (:sync? opts)
+               {go-locked locked
+                <! do}
+               (go-locked
+                store key
+                (or (cache/has? @(:cache store) key)
+                    (<! (-exists? store key opts)))))))
 
 (defn get-in
   "Returns the value stored described by key-vec or nil if the path is
   not resolvable."
   ([store key-vec]
    (get-in store key-vec nil))
   ([store key-vec not-found]
-   (go-locked
-    store (first key-vec)
-    (let [cache         (:cache store)
-          [fkey & rkey] key-vec]
-      (if-let [v (cache/lookup @cache fkey)]
-        (do
-          #_(prn "hitting cache")
-          (swap! cache cache/hit fkey)
-          (clojure.core/get-in v rkey))
-        (let [v (<! (-get store fkey))]
-          #_(prn "getting fkey" fkey)
-          (swap! cache cache/miss fkey v)
-          (clojure.core/get-in v rkey not-found)))))))
+   (get-in store key-vec not-found {:sync? false}))
+  ([store key-vec not-found opts]
+   (trace "get-in on key " key opts)
+   (async+sync (:sync? opts)
+               {go-locked locked
+                <! do}
+               (go-locked
+                store (first key-vec)
+                (let [cache         (:cache store)
+                      [fkey & rkey] key-vec]
+                  (if-let [v (cache/lookup @cache fkey)]
+                    (do
+                      (swap! cache cache/hit fkey)
+                      (if (some? v)
+                        (clojure.core/get-in v rkey not-found)
+                        not-found))
+                    (let [v (<! (-get store fkey opts))]
+                      (swap! cache cache/miss fkey v)
+                      (if (some? v)
+                        (clojure.core/get-in v rkey not-found)
+                        not-found))))))))
```

Comment:
In cases like this where the repetition is local and trivial (in a sense) I prefer to have the code laid out as it is because it requires less mental binding to check that it does the correct thing. In my experience repetition should be avoided if bigger parts repeat and there is a significant risk of introducing mismatches. This is how the rewrite would look like, if you prefer it this way then I can change it, but I do not find this clearer. 

```clojure
               (let [cache         (:cache store)
                      [fkey & rkey] key-vec
                      extract (fn [v]
                                (if (some? v)
                                  (clojure.core/get-in v rkey not-found)
                                  not-found))]
                  (if-let [v (cache/lookup @cache fkey)]
                    (do
                      (swap! cache cache/hit fkey)
                      (extract v))
                    (let [v (<! (-get store fkey opts))]
                      (swap! cache cache/miss fkey v)
                      (extract v))))
```

Reply:
I just realized: you don't need line 57-59 anyway. If `v` is nil, the true branch of the `if-let` is not being executed in the first place. So line 59 will never be executed. 
Or am I overlooking something?